examples/segmentation_3d/assets/algo/algo.py

```diff
@@ -0,0 +1,78 @@
+import monai
+from monai.data import sliding_window_inference
+
+import torch
+
+import substratools as tools
+
+
+class MonaiAlgo(tools.algo.Algo):
+    device = torch.device("cpu")
+    loss_function = monai.losses.DiceLoss(do_sigmoid=True)
+
+    def _get_model(self):
+        model = monai.networks.nets.UNet(
+            dimensions=3,
+            in_channels=1,
+            out_channels=1,
+            channels=(16, 32, 64, 128, 256),
+            strides=(2, 2, 2, 2),
+            num_res_units=2,
+        ).to(self.device)
+        optimizer = torch.optim.Adam(model.parameters(), 1e-3)
+        return (model, optimizer)
+
+    def train(self, X, y, models, rank):
+        # create UNet, DiceLoss and Adam optimizer
+        if not models:
+            model, optimizer = self._get_model()
+        else:
+            model, optimizer = models[0]
+
+        # start a typical PyTorch training
+        epoch_loss = 0
+        step = 0
+        model.train()
+        for inputs, labels in X:
+            step += 1
+            optimizer.zero_grad()
+            outputs = model(inputs)
+            loss = self.loss_function(outputs, labels)
+            loss.backward()
+            optimizer.step()
+            epoch_loss += loss.item()
+            print(f'Step {step}, train_loss:{loss.item()}')
+
+        print(f'Average loss: {epoch_loss / step}')
+
+        return (model, optimizer)
+
+    def predict(self, X, model):
+        y_pred = []
+        model, _ = model  # drop the optimizer
+        model.eval()
+        with torch.no_grad():
+            for inputs, metadata in X:
+                roi_size = (96, 96, 96)
+                sw_batch_size = 4
+                outputs = sliding_window_inference(inputs, roi_size, sw_batch_size, model)
+                y_pred.append((outputs, metadata))
```

Comment:
Could you explain why you are returning both the predicted outputs and the input labels?

(It seems not used inside the score function)

Reply:
For each i-th item of X (`X_i`) I'm returning the prediction `y_pred_i` and the metadata associated with each `X_i`. The metadata contains the name of the file `X_i`comes from which will allow me to then be sure that the `y_i` and `y_pred_i` match the same `X_i`.